mitodl/ocw-studio
# overhaul mass build pipeline

# What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1918

# Description (What does it do?)
This PR adds a replacement for `mass-build-sites.yml`, `mass_build_sites.py`, that uses the `ol-concourse` library to generate a pipeline definition consisting of several batch jobs that build multiple sites in parallel.

# How can this be tested?
 - Make sure you have the latest containers built in docker by running `docker-compose build`
 - Make sure you have the following prerequisites configured:
   - Recent database restore from production
   - Github org for testing set up and configured in your `.env`
   - All sites pushed to your Github org:
```bash
docker-compose exec web ./manage.py reset_sync_states --skip_sync
docker-compose exec web ./manage.py mass_publish draft
```
   - Minio running with credentials in your `.env`
   - Theme assets pipeline pushed up to Concourse with `docker-compose exec web ./manage.py upsert_theme_assets_pipeline`
   - Theme assets pipeline run and completed
   - Concourse `fly` CLI installed with a target pointing to your local Concourse, referred to in these instructions as `local` (https://concourse-ci.org/fly.html#fly-login)
 - Once the above is complete you should have plenty of sites for your mass build pipeline to fetch and build. First, we will generate the pipeline definiton. The pipeline definition will take all of the publishable sites and separate them into batches based on the `OCW_MASS_BUILD_BATCH_SIZE` setting, building a number of sites in each batch simultaneously based on the `OCW_MASS_BUILD_MAX_IN_FLIGHT` setting. The defaults are 20 and 10, respectively, but you can adjust this based on your system specs. On a quad core Dell Latitude 7400 with an i7 and 32GB of RAM, I was able to up the batch size to 40 but your milage may vary, and actually running the pipeline will bog down your system as it is very demanding, so keep that in mind. Run `docker-compose exec web ./manage.py shell` and execute the following:
```python
import json
from urllib.parse import quote
from websites.models import *
from content_sync.pipelines.definitions.concourse.mass_build_sites import (
    MassBuildSitesPipelineDefinitionConfig,
    MassBuildSitesPipelineDefinition
)

version = "draft"
publish_date_field = (
    "publish_date" if version == VERSION_LIVE else "draft_publish_date"
)

# Get all sites, minus any sites that have never been successfully published
sites = Website.objects.exclude(
    Q(**{f"{publish_date_field}__isnull": True}) | Q(url_path__isnull=True)
)
sites = sites.prefetch_related("starter").order_by("name")
pipeline_config = MassBuildSitesPipelineDefinitionConfig(
    sites=sites,
    version=version,
    ocw_studio_url="http://10.1.0.102:8043",
    artifacts_bucket="ol-eng-artifacts",
    site_content_branch="preview",
    ocw_hugo_themes_branch="main",
    ocw_hugo_projects_branch="main",
    offline=False,
    instance_vars=f"?vars={quote(json.dumps({'offline': False, 'prefix': '', 'projects_branch': 'main', 'themes_branch': 'main', 'starter': '', 'version': 'draft'}))}"
)
pipeline_definition = MassBuildSitesPipelineDefinition(config=pipeline_config)
f = open("mass-build-pipeline-test.json", "w")
f.write(pipeline_definition.json(indent=2, by_alias=True))
f.close()
```
 - This will generate a file called `mass-build-pipeline-test.json`. We will push it up to our local Concourse instance by running `fly -t local set-pipeline -p mass-build-sites-test -c mass-build-pipeline-test.json`. Press `y` at the prompt to commit the pipeline.
 - Browse to your local Concourse instance at http://localhost:8080 and login with test / test. You should see the `mass-build-sites-test` pipeline there which should look like a series of vertical bars. These bars are jobs that represent the batches. Click the play icon to unpause the pipeline, then click on the first job and start a build.
 - Watch the first batch run. When it gets to the `across: site` step it should start a number of site builds in parallel based on the `OCW_MASS_BUILD_MAX_IN_FLIGHT` setting. Once the batch is complete, step back to the pipeline overview by clicking the `mass-build-sites-test` breadcrumb in the header.
 - The next batch should trigger automatically. Watch these complete, and verify that it is deploying sites to your local Minio. You can do this by highlighting and copying the `site.s3_path` property and appending it to the end of `http://localhost:8044/`, for example: http://localhost:8044/courses/21w-022-03-writing-and-experience-reading-and-writing-autobiography-spring-2014.
 - You do not need to run the entire mass build to verify this PR, but you should at least make sure some batches complete without error.
